src/azure-cli-core/azure/cli/core/command_recommender.py

```diff
@@ -0,0 +1,196 @@
+# --------------------------------------------------------------------------------------------
+
+import difflib
+
+import azure.cli.core.telemetry as telemetry
+from knack.log import get_logger
+
+logger = get_logger(__name__)
+
+
+class CommandRecommender():
+    """Recommend a command for user when user's command fails.
+    It combines alladin recommendations and examples in help files."""
+
+    def __init__(self, command, parameters, extension, cli_ctx):
+        """
+        :param command: The command name in user's input, which should be corrected if misspelled.
+        :type command: str
+        :param parameters: The parameter arguments in users input.
+        :type parameters: list
+        :param extension: The extension name in user's input if the command comes from an extension.
+        :type extension: str
+        :param cli_ctx: CLI context when parser fails.
+        :type cli_ctx: knack.cli.CLI
+        """
+        self.command = command.strip()
+        self.extension = extension
+        self.cli_ctx = cli_ctx
+
+        self.parameters = self._normalize_parameters(parameters)
+        self.help_examples = []
+        self.aladdin_recommendations = []
+
+    def set_help_examples(self, examples):
+        """Set recommendations from help files"""
+
+        self.help_examples.extend(examples)
+
+    def _set_aladdin_recommendations(self):
+        """Set recommendations from aladdin service.
+        Call the aladdin service API, parse the response and set the recommendations.
+        """
+
+        import json
+        import requests
+        from requests import RequestException
+        from http import HTTPStatus
+        from azure.cli.core import __version__ as version
+
+        api_url = 'https://app.aladdin.microsoft.com/api/v1.0/suggestions'
+        headers = {'Content-Type': 'application/json'}
+        correlation_id = telemetry._session.correlation_id  # pylint: disable=protected-access
+        subscription_id = telemetry._get_azure_subscription_id()  # pylint: disable=protected-access
+
+        context = {
+            'versionNumber': version
+        }
+        if telemetry.is_telemetry_enabled():
+            if correlation_id:
+                context['correlationId'] = correlation_id
+            if subscription_id:
+                context['subscriptionId'] = subscription_id
+
+        parameters = [item for item in self.parameters if item not in ['--debug', '--verbose']]
+        query = {
+            "command": self.command,
+            "parameters": ','.join(parameters)
+        }
+
+        response = None
+        try:
+            response = requests.get(
+                api_url,
+                params={
+                    'query': json.dumps(query),
+                    'clientType': 'AzureCli',
+                    'context': json.dumps(context)
+                },
+                headers=headers,
+                timeout=1)
```

Comment:
I've looked at some of the Azure CLI telemetry for the `az find` command from the past month. In the 90th percentile, it can take ~4.5s to get a response. Internally, the services 90th percentile for the same time period is <400ms, so a lot that is just network latency.

I think a timeout of 5s would cover most users without going exceptionally long. What do you think?

Reply:
As we disscussed, 5s seems too long. 

Anyway, I have recorded the service exceptions into Telemetry. Let's wait for the Telemetry records to see if there are many timeouts there, and then may consider adjust it in next release.